### PR TITLE
Add persistence user configuration options

### PR DESCRIPTION
**Pull request by tomlongridge**
## Goal

Add persistence of set user info (id, email, name) between notifier instantiation.

Cherry-picked version of #469 as some of the merges in this branch had gone astray.

### CODE REVIEW

File: Source/SSKeychain/BSG_SSKeychain.h [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+	SSKeychainErrorDuplicatedItem = errSecDuplicateItem,
+	
+	/** The item cannot be found.*/
+	SSKeychainErrorNotFound = errSecItemNotFound,
+	
+	/** Interaction with the Security Server is not allowed. */
+	SSKeychainErrorInteractionNotAllowed = errSecInteractionNotAllowed,
+	
+	/** Unable to decode the provided data. */
+	SSKeychainErrorFailedToDecode = errSecDecode
```
Comment (kattrali) [inline Source/SSKeychain/BSG_SSKeychain.h]: These values should be prefixed.

File: Source/SSKeychain/BSG_SSKeychain.m [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+NSString *const BSG_kSSKeychainAccountKey = @"acct";
+NSString *const BSG_kSSKeychainCreatedAtKey = @"cdat";
+NSString *const BSG_kSSKeychainClassKey = @"labl";
+NSString *const BSG_kSSKeychainDescriptionKey = @"desc";
+NSString *const BSG_kSSKeychainLabelKey = @"labl";
+NSString *const BSG_kSSKeychainLastModifiedKey = @"mdat";
+NSString *const BSG_kSSKeychainWhereKey = @"svce";
+
+#if __IPHONE_4_0 && TARGET_OS_IPHONE  
+CFTypeRef SSKeychainAccessibilityType = NULL;
```
Comment (kattrali) [inline Source/SSKeychain/BSG_SSKeychain.m]: Missing prefix?

File: Tests/BugsnagSessionTrackerTest.m [hunk 3]
```diff
@@ -63,9 +62,6 @@ - (void)testStartNewAutoCapturedSession {
     XCTAssertNotNil(session);
     XCTAssertNotNil(session.sessionId);
     XCTAssertTrue([[NSDate date] timeIntervalSinceDate:session.startedAt] < 1);
-    XCTAssertNil(session.user.name);
-    XCTAssertNil(session.user.userId);
-    XCTAssertNil(session.user.emailAddress);
```
Comment (kattrali) [inline Tests/BugsnagSessionTrackerTest.m]: Do these have sensible values now or something? I'd expect that they would still be nil.

File: UPGRADING.md [hunk 4]
```diff
@@ -24,6 +24,9 @@ Swift:
 
 + BSGConfigurationErrorDomain
 + BSGConfigurationErrorCode
++ config.persistUser
++ [config persistUserData]
++ [config deletePersistedUserData]
```
Comment (kattrali) [inline UPGRADING.md]: What is the end user intent for accessing these? If there isn't one, then they shouldn't be public on config.

File: Source/BugsnagConfiguration.h [hunk 5]
```diff
@@ -42,6 +42,14 @@ typedef NS_ENUM(NSInteger, BSGConfigurationErrorCode) {
     BSGConfigurationErrorInvalidApiKey = 0
 };
 
+/**
+ * User info persistence keys
+ */
+extern NSString * _Nonnull const kBugsnagUserKeychainAccount;
+extern NSString * _Nonnull const kBugsnagUserEmailAddress;
+extern NSString * _Nonnull const kBugsnagUserName;
+extern NSString * _Nonnull const kBugsnagUserUserId;
```
Comment (kattrali) [inline Source/BugsnagConfiguration.h]: These shouldn't be public if they aren't intended for public use.

File: Source/SSKeychain/BSG_SSKeychain.h (hunk 1 above)
Comment (robinmacharg) [reply Source/SSKeychain/BSG_SSKeychain.h]: My bad.  Fixed.

File: Tests/BugsnagSessionTrackerTest.m (hunk 3 above)
Comment (robinmacharg) [reply Tests/BugsnagSessionTrackerTest.m]: Persisted user details caused these to fail.  These are now cleared out in setup(). The AssertNils have been restored.

File: UPGRADING.md (hunk 4 above)
Comment (robinmacharg) [reply UPGRADING.md]: Fair point.  Removed and appropriate @interfaces added to tests.

**Pull request merged**